### PR TITLE
[Messenger] use Envelope internally, return void, add EnvelopeHandlerInterface and other cleanups

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 4.2
| Bug fix?      | yes
| New feature?  | yes
| BC breaks?    | yes
| Deprecations? | no
| Tests pass?   | not yet
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

Sorry in advance for this late PR, I've just reviewed the component for the first time and I think we should do some changes in its internal design, so here we are, last chance before tagging it as stable.

Each item in this list can potentially trigger a discussion. Here are the topics of this PR:
- make everything *internally* deal with `Envelope` objects (middlewares, transports, etc.)
- ~replace `EnvelopeAwareInterface` (for middlewares) by `EnvelopeHandlerInterface` (for userland handlers)~ [edit: bad idea, see discussion]
- remove the return value of `dispatch()`+`handle()`: we never use it and it contradicts the very idea of async (mutable messages can be used instead, but that's not our business)
- added `string $name = null` as 2nd argument to dispatch(), used as dispatching key when provided for more flexibility (it's not always possible/desirable to create a new class to define a dispatching key)
- ~locating senders/handlers using the parent classes+interfaces of the message is removed.~ To me, locating based on the type hierarchy is a code smell that we hijack the type hierarchy to make it fit what should be done via configuration. [edit: now kept to reduce the impact of this PR - could be deprecated later if this smell is proven by experience in the future]
- remove `AllowNoHandlerMiddleware` and replace it with a constructor argument on `HandleMessageMiddleware`: using exceptions for signaling is going to slow down the component for no real reasons to me. This solves it. 
- rename some methods / change some interfaces for consistency/ease of use

 I expect this to raise some discussions. To resolve them one by one, I propose to split this PR into separate ones while we move our common understanding forward.

Meanwhile, I didn't update tests nor DI integration. I'll do as this moves forward.

### CODE REVIEW

File: src/Symfony/Component/Messenger/Handler/EnvelopeHandlerInterface.php [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ */
+
+namespace Symfony\Component\Messenger\Handler;
+
+/**
+ * A marker interface for handlers that want to receive messages wrapped in their envelopes.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+interface EnvelopeHandlerInterface extends MessageHandlerInterface
```
Comment (ro0NL) [inline src/Symfony/Component/Messenger/Handler/EnvelopeHandlerInterface.php]: :+1:

File: src/Symfony/Component/Messenger/Middleware/MiddlewareInterface.php [hunk 2]
```diff
@@ -11,15 +11,12 @@
...
  */
 interface MiddlewareInterface
 {
-    /**
-     * @param object $message
-     *
-     * @return mixed
-     */
-    public function handle($message, callable $next);
+    public function handle(Envelope $envelope, callable $next): void;
```
Comment (ro0NL) [inline src/Symfony/Component/Messenger/Middleware/MiddlewareInterface.php]: :+1:

**Comment by Koc** [comment]
> remove the return value of dispatch()+handle(): we never use it and it contradicts the very idea of async (mutable messages can be used instead, but that's not our business)

what about query bus? It would be nice provide query bus out of the box

File: src/Symfony/Component/Messenger/MessageBusInterface.php [hunk 3]
```diff
@@ -19,11 +19,7 @@
     /**
      * Dispatches the given message.
      *
-     * The bus can return a value coming from handlers, but is not required to do so.
-     *
      * @param object|Envelope $message The message or the message pre-wrapped in an envelope
```
Comment (Koc) [inline src/Symfony/Component/Messenger/MessageBusInterface.php]: What about adding phpdoc for the new argument? Describe when this name can be useful

**Comment by nicolas-grekas** [comment]
> what about query bus? It would be nice provide query bus out of the box

They could use mutable command objects (e.g. a `setResult()`). This would benefit the type system since the argument to this method could be type-hinted.

Right now, the return value blurs the scope of the component to me. Also look at e.g. ChainHandler: returning the array of all results makes no sense: this is configuration leaking to the outside to me.
Makes sense?

**Comment by nicolas-grekas** [comment]
> Meanwhile, I'm not convinced EnvelopeHandlerInterface would be useful.

Since `dispatch()` is not an internal detail, it means ppl that use it to wrap their messages should be able to receive them also. Otherwise, why allow dispatching envelope in the first place? To me, either we make Envelope purely internal (ie never dispatch them) - or handlers need to be able to receive them, in an opt-in way.

**Comment by nicolas-grekas** [comment]
Note that I may be missing the reason why #27322 has been rejected: right now, I'm thinking about middlewares as things that are tightly bound to Messenger so that they legitimately know about it, and about Envelope especially.
The boundary that is loosely coupled is the caller of `dispatch()`, and it's corresponding handlers. *Those* don't need to care about Envelope.
So, @sroze maybe: what are the use cases you envision where middlewares are on the outer boundary of the component?

**Comment by ogizanagi** [comment]
> Since Envelope is not an internal detail, [...] handlers need to be able to receive them, in an opt-in way.

It's not about the envelope being internal or not regarding this. Envelope are meant for runtime "config" of middleware & transports. If the handler need some data to treat the message, it should be part of it, not the envelope.

**Comment by nicolas-grekas** [comment]
> If the handler needs some data to treat the message, it should be part of it, not the envelope.

aren't there any cases where the handler will want to know about the way *routing* did happen? E.g. number of retries, timestamps, some routing/logging identifier, etc. - Thinks that could be ofinterest for some handlers?

**Comment by ogizanagi** [comment]
Obviously not to me. Handlers are pure business logic. The use-cases you mention are valid as a middleware (which if really needed can mutate the message, but those are usually supposed immutable).

File: src/Symfony/Component/Messenger/Handler/ChainHandler.php [hunk 4]
```diff
@@ -30,21 +31,17 @@ class ChainHandler
...
     }
 
-    public function __invoke($message)
+    public function __invoke(Envelope $envelope)
     {
-        $results = array();
-
         foreach ($this->handlers as $handler) {
-            $results[] = $handler($message);
+            $handler($envelope instanceof EnvelopeHandlerInterface ? $envelope : $envelope->getMessage());
```
Comment (ro0NL) [inline src/Symfony/Component/Messenger/Handler/ChainHandler.php]: `$handler instanceof`

**Comment by nicolas-grekas** [comment]
Closing here because this has been split into several PRs already - or will soon be for the remaining ideas.
See #28914, #28911, #28909, #28908, #28907

**Pull request closed without merging**